orangecontrib/text/twitter.py

```diff
@@ -0,0 +1,121 @@
+                                        callback=self.callback)
+
+    def check_credentials(self):
+        try:
+            self.auth.get_authorization_url()
+        except tweepy.TweepError:
+            return False
+
+        return True
+
+    def construct_api(self):
+        return tweepy.API(self.auth)
+
+    def __eq__(self, other):
+        return self.consumer_key == other.consumer_key \
+               and self.consumer_secret == other.consumer_secret
+
+
+class TwitterAPI:
+    """Fetch tweets from the Tweeter API"""
+
+    string_fields = [
+        'text', 'created_at', 'lang'
+    ]
+
+    continuous_fields = [
+        'favorite_count', 'retweet_count'
+    ]
+
+    discrete_fields = [
+
+    ]
+    supported_fields = string_fields + continuous_fields + discrete_fields
+
+    def __init__(self, credentials):
+        """
+
+        Args:
+            credentials (Credentials):
+        """
+        self.key = credentials
+        self.api = credentials.construct_api()
+        self.stream = tweepy.Stream(auth=self.api.auth, listener=Listener(self))
+        self.statuses_lock = Lock()
+
+        self.container = []
+
+    # TODO: add arguments
+    def search(self, *, word_list, max_tweets, lang=None):
+        if not word_list:
+            raise ValueError("Word list can't be empty.")
+
+        query = ' OR '.join(word_list)
+
+        for status in tweepy.Cursor(self.api.search, q=query, lang=lang).items(max_tweets):
+            self.add_status(status)
+
+    def listen(self, *, word_list):
+        self.stream.filter(track=word_list, async=True)
+
+    def add_status(self, status):
+        status_record = {field: getattr(status, field)
+                         for field in self.supported_fields}
+        self.statuses_lock.acquire()
+        self.container.append(status_record)
+        self.statuses_lock.release()
+        self.on_status()
+
+    def create_corpus(self):
+        """Creates corpus from collected tweets.
+        """
+        X = np.array([
+            [float(record[field]) for field in self.continuous_fields]
+            for record in self.container
+        ])
+
+        metas = np.array([
+             [record[field] for field in self.string_fields]
+             for record in self.container
+        ], dtype=object)
+
+        metas_attr = [data.StringVariable(name) for name in self.string_fields]
+        domain = data.Domain(attributes=[data.ContinuousVariable(name) for name in self.continuous_fields],
+                             metas=metas_attr)
+
+        return Corpus(X=X, metas=metas, domain=domain, text_features=metas_attr)
+
+    def on_status(self):
+        pass
+
+
+class Listener(tweepy.StreamListener):
+    def __init__(self, api):
+        super().__init__()
+        self.api = api
+
+    def on_status(self, status):
+        self.api.add_status(status)
+
+    def on_error(self, status):
+        print(status)
```

Comment:
This print should be removed in the final version. If there is an error it should be propagated to widget instead of printing into console.